[WIP] plugin: add support for updating bank of a pending job

cmoussa1 opened the pull request:
~~note: this is built on top of #399, so I think it might be best to hold off on this PR until that one lands.~~ (#399 has now landed)

#### Problem

flux-accounting currently has no support for a user/bank to update the bank they want their pending job to run under.

---

This is a PR that adds support for updating the bank of a pending job. It adds a new callback to the plugin which looks for the `"job.update.attributes.system.bank"` topic string and validates the new bank that the user is trying to update their job to. If they have access to the bank they want to update their job to _but_ this new bank is currently at its max running jobs limit, the plugin will reject the update. Since `job.state.depend` is the final place where a job can have dependencies added to it (and adding/removing dependencies is how the plugin handles running job limits), I've chosen to just reject the update and keep the job under its current bank.

Once a new bank has been validated, its attributes are then applied to the job's `bank_info` struct in the `job.update` callback. You'll notice that before the new attributes/limits are assigned to the job, the `active_jobs` count of the old bank is decremented. Then, just before the new attributes/limits are assigned to the job, the `active_jobs` count of the new bank is incremented (since this count usually gets incremented when the job is in `job.new`).

A new test file, `t1031-mf-priority-update-bank.t` is also added to the test suite which simulates submitting some jobs under default and non-default banks and updating the bank between each, while also making sure that a job being updated to a bank that is already at their current max running jobs limit gets rejected.

---

#### TODO

- [x] add a check in the test suite to ensure that active/running jobs limits are as expected after an update

cmoussa1 committed:
plugin: add bank name to bank_info struct

Problem: the bank_info struct does not contain the name of the bank in
the struct, which can make it more tedious to find the name of the bank
because you have to look at the key of the map in order to find it. It
would be more convenient if the name was also in the struct.

Add the bank name to the struct attached with each user/bank
combination.

cmoussa1 committed:
plugin: add job.update for banks

Problem: flux-accounting currently has no support for a user to update
the bank of their job.

Add a new callback to the plugin which looks for the
"job.update.attributes.system.bank" topic string and validates the new
bank that the user is trying to update their job to. If they have access
to the bank they want to update their job to but it is currently at its
max active jobs OR max running jobs limits, reject the update.

Add a check for this update in the job.update callback which will look
for an updated bank. If the updated bank has been validated, change the
bank_info struct assigned to that job to reflect the attributes of the
new bank.

cmoussa1 committed:
t: add tests for updating bank

Problem: flux-accounting has no tests for updating the bank of a pending
job.

Add some tests.

grondo commented on src/plugins/mf_priority.cpp:
```diff
@@ -1078,8 +1078,8 @@ static int run_cb (flux_plugin_t *p,
 /*
- *  apply an update on a job with regard to its queue once it has been
- *  validated.
+ *  apply an update on a job with regard to its queue or associated bank once
```
Commit message: " flux-accounting currently has no support for a user/bank to update the bank they want their pending job to run under." Should this just be "flux-accounting currently has no support for a user to update the bank of their job"?

cmoussa1 replied:
That's a better message than what I had suggested 😅 thanks! I've just updated this commit message.

grondo reviewed (commented):
Just did a quick first pass here. I wonder if some tests that update both bank and queue at the same time should be added, both for valid and invalid combinations of updates?

Also, as an aside, while reviewing this I was thinking that it might be good to start thinking about refactoring the bank_info structure in his plugin into a class. The class interface could be pulled out o the plugin itself, and tested with unit tests. I think this would really clean up some of the code for looking up and inserting new bank information for users, and make the code easier to read (I can open an issue on this if you'd like)

cmoussa1 committed:
t: add tests for updating both queue and bank

Problem: flux-accounting has no tests for updating both the queue and
the bank of a job at the same time.

Add a new test file which tests updating both the queue and the bank at
the same time as well as some combinations where at least one attribute
update is expected to fail (i.e an invalid bank or an invalid queue).

In both of the cases where at least one attribute fails, the entire
request will fail and the job will remain under its original attributes.

cmoussa1 commented:
> Just did a quick first pass here. I wonder if some tests that update both bank and queue at the same time should be added, both for valid and invalid combinations of updates?

That's a good suggestion, thanks. I've added a commit in this PR that adds another test file that tests updating both queue and bank at the same time as well as a couple of invalid combinations (i.e the queue specified or the bank specified was invalid). Let me know if this is what you had in mind!

> Also, as an aside, while reviewing this I was thinking that it might be good to start thinking about refactoring the bank_info structure in his plugin into a class. The class interface could be pulled out o the plugin itself, and tested with unit tests. I think this would really clean up some of the code for looking up and inserting new bank information for users, and make the code easier to read (I can open an issue on this if you'd like)

Oh sure! That could be really helpful. Yes, if you'd like to open an issue I think that would be useful for me. :-)

cmoussa1 commented:
Now that the priority plugin is cleaned up and there are new external functions, I plan to revisit this and update it to reflect the new changes. Will convert this back to [WIP] in the meantime!

cmoussa1 commented:
Going to close this as I now have a (hopefully much cleaner) re-implementation of the changes proposed in this PR.
